boa/src/builtins/string/mod.rs

```diff
@@ -479,39 +482,90 @@ impl String {
-
-                    if val.find("$$").is_some() {
-                        result = val.replace("$$", "$")
-                    }
-
-                    if val.find("$`").is_some() {
-                        let start_of_match = mat.start();
-                        let slice = &primitive_val[..start_of_match];
-                        result = val.replace("$`", slice);
-                    }
-
-                    if val.find("$'").is_some() {
-                        let end_of_match = mat.end();
-                        let slice = &primitive_val[end_of_match..];
-                        result = val.replace("$'", slice);
-                    }
-
-                    if val.find("$&").is_some() {
-                        // get matched value
-                        let matched = caps.get(0).expect("cannot get matched value");
-                        result = val.replace("$&", matched.as_str());
-                    }
-
-                    // Capture $1, $2, $3 etc
-                    if re.is_match(&result) {
-                        let mat_caps = re.captures(&result).unwrap();
-                        let group_str = mat_caps.get(1).unwrap().as_str();
-                        let group_int = group_str.parse::<usize>().unwrap();
-                        result = re
-                            .replace(result.as_str(), caps.get(group_int).unwrap().as_str())
-                            .to_string()
+                    let mut result = StdString::new();
+                    let mut chars = val.chars().peekable();
+
+                    let m = caps.len();
+
+                    while let Some(first) = chars.next() {
+                        if first == '$' {
+                            let second = chars.next();
+                            let second_is_digit = second.map(|ch| ch.is_digit(10)).unwrap_or(false);
+                            // we use peek so that it is still in the iterator if not used
+                            let third = if second_is_digit { chars.peek() } else { None };
+                            let third_is_digit = third.map(|ch| ch.is_digit(10)).unwrap_or(false);
+
+                            match (second, third) {
+                                (Some('$'), _) => {
+                                    // $$
+                                    result.push('$');
+                                }
+                                (Some('&'), _) => {
+                                    // $&
+                                    let matched = caps.get(0).expect("cannot get matched value");
+                                    result.push_str(matched.as_str());
+                                }
+                                (Some('`'), _) => {
+                                    // $`
+                                    let start_of_match = mat.start();
+                                    let slice = &primitive_val[..start_of_match];
+                                    result.push_str(slice);
```

Comment:
```suggestion
                                    result.push_str(&primitive_val);
```
This should work since `RcString` implements `Borrow` with type `&str` otherwise,
Calling `.as_str()` on `RcString` returns a `&str` this makes the code more readable